src/main/webapp/app/overview/course-discussion/course-discussion.component.ts

```diff
@@ -0,0 +1,271 @@
+import { map, Subscription } from 'rxjs';
+import { Course } from 'app/entities/course.model';
+import { Exercise } from 'app/entities/exercise.model';
+import { Lecture } from 'app/entities/lecture.model';
+import { MetisService } from 'app/shared/metis/metis.service';
+import { Post } from 'app/entities/metis/post.model';
+import { FormBuilder, FormGroup } from '@angular/forms';
+import { Reaction } from 'app/entities/metis/reaction.model';
+import { ButtonType } from 'app/shared/components/button.component';
+import { HttpResponse } from '@angular/common/http';
+import { CourseManagementService } from 'app/course/manage/course-management.service';
+
+interface ContextFilterOption {
+    courseId?: number;
+    lectureId?: number;
+    exerciseId?: number;
+    courseWideContext?: CourseWideContext;
+}
+
+interface ContentFilterOption {
+    searchText?: string;
+}
+
+@Component({
+    selector: 'jhi-course-discussion',
+    templateUrl: './course-discussion.component.html',
+    styleUrls: ['./course-discussion.scss'],
+    providers: [MetisService],
+})
+export class CourseDiscussionComponent implements OnInit, OnDestroy {
+    course?: Course;
+    exercises?: Exercise[];
+    lectures?: Lecture[];
+    currentPostContextFilter: ContextFilterOption;
+    currentSortCriterion = PostSortCriterion.CREATION_DATE;
+    currentSortDirection = SortDirection.DESC;
+    currentPostContentFilter: ContentFilterOption;
+    searchText?: string;
+    formGroup: FormGroup;
+    createdPost: Post;
+    posts: Post[];
+    isLoading = true;
+    readonly CourseWideContext = CourseWideContext;
+    readonly SortBy = PostSortCriterion;
+    readonly SortDirection = SortDirection;
+    readonly PageType = PageType;
+    readonly ButtonType = ButtonType;
+    readonly pageType = PageType.OVERVIEW;
+
+    private postsSubscription: Subscription;
+    private paramSubscription: Subscription;
+
+    constructor(
+        protected metisService: MetisService,
+        private activatedRoute: ActivatedRoute,
+        private courseManagementService: CourseManagementService,
+        private formBuilder: FormBuilder,
+    ) {}
+
+    /**
+     * on initialization: initializes the metis service, fetches the posts for the course, resets all user inputs and selects to defaults,
+     * creates the subscription to posts to stay updated on any changes of posts in this course
+     */
+    ngOnInit(): void {
+        this.paramSubscription = this.activatedRoute.parent!.params.subscribe((params) => {
+            const courseId = parseInt(params['courseId'], 10);
+            this.courseManagementService.findOneForDashboard(courseId).subscribe((res: HttpResponse<Course>) => {
+                if (res.body !== undefined) {
+                    this.course = res.body!;
+                    this.lectures = this.course?.lectures;
+                    this.exercises = this.course?.exercises;
+                    this.metisService.setCourse(this.course!);
```

Comment:
I am not sure about this dataflow. From what I can see, the metisService gets used to hold a state, e.g. the course, with some components setting it and others getting it. Did you consider to use a solution where the data is instead transferred using Angular inputs and outputs? Making the service stateful seems a bit dangerous, if any component fails to update the state, the whole thing breaks. I am not knowledgeable enought to understand the details, it just seems a bit strange to have a stateful service.

Reply:
There are indeed other stateful services, such as the `CourseScoreCalculationService`. In this case, I found this to be the most maintainable and understandable solution. The `metisService` is only provided in metis-related downstream components that rely on it - not in root. Those components are nested up to a depth of ~4, so handling with `@Input` and `@Output` is pretty messy and I decided that the `metisService` provides the posts and tags as Observables where all interested components can listen to, whereas the course is set on service initialization from the nearest components, where the service is injected, i.e. `PageDiscussionComponent` and `CourseDiscussionComponent`. Within these components I made sure to actually set the course on component initialization and render the template only `*ngIf="course"`. So I'd argue, that this is a reasonable implementation.

Reply:
Thanks for the explanation, under those circumstances I agree this is the most reasonable design :+1: